### ISSUE TITLE
WO-018: training harness

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#23]
> **WO-018** - Phase 1 - Owner **MID-strong** - Difficulty **3/5** - Gate **G2**
> Card: [`workorders/WO-018.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-018.md)

## What this is

This is the machinery that actually runs a training job: a vectorised batch of environments stepped
in lockstep, the rollout loop that feeds the WO-017 adapter, the entropy anneal, checkpoints,
periodic evaluation with the ledger attached, wall-clock logging, and the run manifest. Both Phase-1
experiments (WO-019's `N = 1` recovery check and WO-020's `N = 20` gate) drive this harness; it never
drives them.

Two disciplines run through it. The reward the environment returns is already the exact CONTRACT
rule-4 reward, multiplied by the analytic scale - the harness stores it, computes GAE from it, and
hands it to the reference PPO without touching it, and no wrapper it applies may rescale it either.
And `TrainConfig`'s sizing fields deliberately have no defaults, because the batch shape and the
training budget are experimental choices belonging to the card that commissions the run; inventing a
default here would make an unrecorded choice look like a specification.

## Depends on

- `WO-017`
- `WO-011`

**Blocked by:** #22 (WO-017), #16 (WO-011)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` - sections 6.1 (`IPPO`), 2.5 (agent-steps per period), 2.12 (geometric episodes), 2.15
  and 4.3 (common random numbers by `seed_env`), 4.1 rows 1 and 4 (THE ONLY ROWS PERIODIC EVALUATION
  MAY TOUCH) plus the held-out paragraph, 4.4 (measurement window), 4.5 (the G2 criteria this harness
  produces runs for), 12.3 (this card), 14 (compute budget). `PLAN.md` is git-ignored and never
  committed; it must be present at the repository root locally, supplied out of band.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4, 6, 8, 9,
  10, plus 2, 3, 12
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - v1.0.0, the
  frozen environment and agent surfaces
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig`, `hash`, `steps_per_period`, `tenure`, `min_periods`, `max_periods`
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py) - the
  registry behind those defaults
- [`gosplan/env/env.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/env.py) -
  `GosplanEnv`, reset/step, `attach_ledger`, `record_step`
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  `StepInfo`
- [`gosplan/agents/ppo/adapter.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/ppo/adapter.py)
  - `IPPO`, `PPOConfig`, `manifest_entry`, the checkpoint methods
- [`gosplan/agents/ppo/train.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/ppo/train.py)
  - WRITE TARGET. `TrainConfig` is declared (sizing fields with NO defaults); seven bodies missing
- [`gosplan/metrics/ledger.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/ledger.py)
  - `Ledger`, `StepRecord`, `bound_binding`, `write_manifest`, `MANIFEST_FIELDS`
- [`gosplan/metrics/phenomena.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/phenomena.py)
  - `phenomenon_bunching`, `phenomenon_padding` - ROWS 1 AND 4 ONLY
- [`gosplan/metrics/__init__.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/__init__.py)
  - `resolve_estimators`, the backend and version for the manifest
- the pinned reference PPO's source at the version WO-017 recorded (read-only, outside this repository)
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312
- [`Makefile`](https://github.com/Xocas12/gosplan-env/blob/main/Makefile) - the targets the completion
  command mirrors

Nothing else. Not `gosplan/experiments/` - WO-019 and WO-020 drive this harness, not the reverse.
Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/agents/ppo/train.py` - `make_env_batch`, `entropy_coefficient`, `evaluate`,
  `checkpoint_path`, `log_update`, `train_manifest_extra`, `train`. `TrainConfig` keeps its declared
  fields, and its sizing fields keep having NO defaults

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] smoke test - PLAN section 12.3 says only "Must pass: smoke test (100 updates on `N = 1`)" for
      this card and names no path. By the naming of the one smoke test PLAN does name
      (`tests/unit/test_mc_sanity_runs.py`, WO-012) the expected path is
      `tests/unit/test_train_smoke.py`: 100 updates at `N = 1` complete and the run writes a manifest
      carrying every CONTRACT rule-10 field. IT IS NOT ON WO-002'S WRITE LIST, so it is expected to
      be ABSENT. If it is absent, FILE AN AMBIGUITY REPORT asking the lead to author it (only the
      lead writes frozen tests, rule 2), say so in the report, and DO NOT WRITE IT YOURSELF.
- [ ] the full frozen suite - `uv run pytest -q`, green

```bash
uv run ruff check gosplan/agents/ppo/train.py
uv run ruff format --check gosplan/agents/ppo/train.py
uv run pytest tests/unit/test_train_smoke.py -q
uv run pytest -q
```

The third line runs only if the frozen suite contains that file (see "Must pass"); otherwise report
its absence.

## Implementation notes

What the card requires, and therefore what this module contains: a VECTORISED ENVIRONMENT BATCH,
CHECKPOINTS, PERIODIC EVALUATION WITH THE LEDGER, THE MANIFEST, CRN BY `seed_env`, WALL-CLOCK
LOGGING, and the ENTROPY ANNEAL `0.01 -> 0.001`.

**1. `TrainConfig` has no sizing defaults, on purpose.** PLAN fixes the TECH hyper-parameters
(`gamma`, `lambda_GAE`, the learning rate, the clip coefficient, the entropy endpoints - all in
`PPOConfig`) but not the batch shape or the training budget; those follow the experiment card that
commissions the run (WO-019's `N = 1` comparison, WO-020's `N = 20` gate) inside the PLAN section 14
envelope. Inventing a default here would make an unrecorded experimental choice look like a
specification (CONTRACT rule 3). This is the single most likely place to break rule 3 on this card.

**2. Batch and seeds.** `make_env_batch(cfg, n_envs, seed_env)` builds `n_envs` copies of
`GosplanEnv` stepped in lockstep; THE BATCH AXIS IS OUTER, THE `N` ENTERPRISES INNER. Seeds derive
from `seed_env`, so arms share environment draws by construction (PLAN sections 2.15 and 4.3), and
`seed_policy` stays a separate stream (CONTRACT rule 9). EVALUATION USES ITS OWN SEED BLOCK, so it
never consumes the training sequence.

**3. Updates and the anneal.**

```
n_updates = total_agent_steps // (n_envs * rollout_steps)
entropy_coefficient(update, n_updates, ppo_cfg)
    interpolates entropy_coef_start -> entropy_coef_end  (0.01 -> 0.001)
    over exactly that count
```

An agent-step is one PRODUCE or REPORT step, so a period is `cfg.incentive.steps_per_period + 1` of
them. A rollout that is not a whole number of periods is LEGAL - the update boundary may fall inside
a period - and the manifest records it.

**4. Periodic evaluation.** `evaluate(...)` runs `eval_episodes` episodes with the ledger attached and
computes ROWS 1 (bunching) AND 4 (padding) ONLY, through WO-016's `phenomenon_bunching` and
`phenomenon_padding`, over the PLAN section 4.4 window: `t >= 2`, no end-of-episode exclusion under
geometric termination, at-bound reports included and flagged. CHECKPOINT SELECTION MAY NOT BE DRIVEN
BY `welfare_true` OR `val_measured` (CONTRACT rule 6). They may be plotted after the fact from the
ledger; they may not be evaluation targets used to select a checkpoint.

**5. Manifest (CONTRACT rule 10).** `train_manifest_extra` supplies `reference_ppo_version` (from
`IPPO.manifest_entry`), `estimator_backend` and `estimator_version` (from `resolve_estimators`),
`bunching_settings` (from `phenomena.py`), the seeds, the flags and the `TrainConfig` itself. Every
field of `MANIFEST_FIELDS` is present; inapplicable ones are `null`.

**6. No reward touching.** The environment returns the rule-4 reward, ALREADY multiplied by the
analytic `reward_scale(cfg)`. Store it, compute GAE from it, hand it to the reference PPO. Do not
normalise, clip, shape or rescale it, and DO NOT WRAP THE ENVIRONMENT IN ANYTHING THAT RESCALES IT.
Rule 4 is a property of the WHOLE STACK, not only of the adapter: any vector-environment wrapper this
harness applies is checked for a reward normaliser too, and the report must state explicitly which
wrappers are applied and that none of them normalises reward.

**7. No held-out phenomenon.** Rows 2, 5, 6 and 7 of PLAN section 4.1 are held out: no plot, table or
test of them before the Phase-2 acceptance run - not in a learning curve, not in a debug print.

**8. No silent bound change.** If the ledger raises `BOUND_BINDING` (more than 1% of reports at
`rho_max`), the flag rides in the manifest and every report of the run displays it. The bound is
NEVER widened to clear it (CONTRACT rule 8).

**9. Checkpoints and logging.** `runs/<env_cfg.hash()>/checkpoints/` via `checkpoint_path`, every
`checkpoint_every_updates`. `log_update` appends one row per update with the wall clock, so the PLAN
section 14 throughput estimate (about 3k steps/s in NumPy) can be checked against reality. Report the
observed agent-steps per second.

## Forbidden

- Any reward normalisation, clipping, shaping or rescaling, in the harness or in any wrapper it
  applies (rule 4).
- Feeding `StepInfo`, `State`, `PlannerView`, `welfare_true` or `val_measured` to the learner, or
  selecting a checkpoint by any of them (rule 6).
- Computing, plotting or logging any PLAN section 4.1 row other than 1 and 4.
- Inventing a default for any sizing field of `TrainConfig`, or changing a `PPOConfig` TECH value.
- Widening `report_max_ratio` or suppressing `BOUND_BINDING` (rule 8).
- Implementing or forking PPO here - the learner is the pinned reference the adapter wraps.
- Editing `gosplan/agents/ppo/adapter.py`, `gosplan/env/`, `gosplan/metrics/`, `tests/`, `ref/`,
  `spec/spec.py` or `PLAN.md`. Running `tests/acceptance/` (rule 13).

## Contract rules that bite here

- **Rule 3 (STOP AND REPORT).** Above all: DO NOT INVENT A DEFAULT FOR ANY SIZING FIELD OF
  `TrainConfig`. Choosing "the reasonable default" is a violation: file an AMBIGUITY REPORT
  (`workorders/AMBIGUITY_TEMPLATE.md`) and end the session.
- **Rule 4 (reward terms).** A property of the whole stack: no wrapper this harness applies may
  normalise, clip, shape or rescale reward. Per-batch advantage normalisation inside PPO is
  permitted.
- **Rule 6 (welfare blindness).** `StepInfo` is consumed here ONLY to append `StepRecord`s and raise
  run flags. It never enters an observation, an advantage, a loss, or any logged quantity that feeds
  back into training, and it never selects a checkpoint.
- **Rule 2 (frozen tests).** Only the lead writes frozen tests. If `tests/unit/test_train_smoke.py`
  is absent, ask for it; do not author it.
- **Rule 8 (bounds are results).** `BOUND_BINDING` rides in the manifest and every report; the bound
  is never widened to clear it.
- **Rule 9 (RNG).** CRN by `seed_env`; `seed_policy` is a separate stream; evaluation uses its own
  seed block.
- **Rule 10 (manifest).** Every run writes a manifest with every field of `MANIFEST_FIELDS`.
- **Rule 12 (whitelist).** Read only the listed files, write only `gosplan/agents/ppo/train.py`, run
  the completion command verbatim, report in the format above.
- **Rule 13 (tests are not experiments).** Do not run `tests/acceptance/`.

## Definition of done

- [ ] The smoke test (100 updates at `N = 1`) green, or its absence reported and an AMBIGUITY REPORT
      filed asking the lead to author it
- [ ] Full frozen suite (`uv run pytest -q`) green
- [ ] `uv run ruff check . && uv run ruff format --check .` clean
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] `TrainConfig`'s sizing fields still have NO defaults
- [ ] Observed agent-steps per second recorded against the PLAN section 14 estimate, and the manifest
      written carries every `MANIFEST_FIELDS` entry
- [ ] No file outside "Write only" touched - expect exactly `gosplan/agents/ppo/train.py`
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: (1) tests passed/failed with output - the smoke
      test or its absence and the full frozen suite verbatim, plus the observed agent-steps per
      second and the manifest written; (2) ambiguity reports filed, or "none" - the smoke-test path
      is the likely one; (3) files changed; (4) anything you were tempted to invent and did not -
      especially any `TrainConfig` default, any wrapper you considered adding, and any evaluation
      metric beyond rows 1 and 4, with an explicit statement of which vector-environment wrappers are
      applied and that none of them normalises reward
